---
title: "Coverage, depth and nucleotide transitions in aDNA using queries and reads"
author: "Tomos Prys-Jones"
date: "12/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(cluster)    
library(dplyr)
library(tidyverse) 
library(stats)
library(rlang)
library(reshape)
library(ggplot2)
library(plotly)
library(plot3D)
library(randomcoloR)
library(scales)
library("gridExtra")   
```


```{r}
rm(list = ls())
if(!is.null(dev.list())) dev.off()
# Change the directory below to the host of interest:
base_dir <- "../data/coverage_stats/a_bison/"
file_list <- list.files(base_dir, pattern = "coverage_stats.txt$")
readfile_list_yn <- str_detect(file_list, "reads_coverage_stats.txt")
readfile_list <- file_list[readfile_list_yn == T]
queryfile_list <- file_list[readfile_list_yn == F]
```


################################################################
#### Section A: QUERY DATAFRAME - COVERAGE AND DEPTH STATS #####
################################################################

The query depth and coverage stats were generated by:
  - Extracting queries from MTSv.
  - And aligned against a reference genome (where available) using: BWA aln.
  - SAMTOOLS was used to calculate the depth and coverage stats in each _reads_coverage_stats.txt file.

```{r, message=FALSE}
# Empty df that will contain stats from each of the separate stats.txt files:
coverage_depth_dataframe <- data.frame(sampleid = character(),
                                       taxid = integer(), 
                                       rname = character(),
                                       startpos = numeric(),
                                       endpos = numeric(),
                                       numquery = numeric(),
                                       covbases = numeric(),
                                       coverage = numeric(),
                                       meandepth = numeric(),
                                       meanbaseq = numeric(),
                                       meanmapq = numeric())   
# For loop below goes through each of the _coverage_stats.txt files and adds hem to the df above.
# In many cases there are multiple sequences per reference, resulting in multiple lines per _coverage_stats.txt file
# Empty df below: will contain the summary statistics from the _coverage_stats.txt files, 
# i.e. it will have a single line for each set of taxon queries that were extracted by MTSv 
coverage_depth_dataframe_queries_reduced <- data.frame(taxid = numeric(), 
                                                       sampleid = character(),
                                                       scaled_depth = numeric(),
                                                       scaled_coverage = numeric(),
                                                       queryno_scaled_by_ref = numeric(),
                                                       taxid_sampleid = character())
for(i in 1:length(queryfile_list)){
  #i <- 1
  focal_filename <- queryfile_list[i]
  focal_dir_filename <- paste0(base_dir, focal_filename)
  focal_file <- read_tsv(focal_dir_filename, col_names = TRUE)
  focal_file <- as.data.frame(focal_file)
  taxid <- sub(".*taxid_", "", focal_filename)
  taxid <- sub("_coverage_stats.txt.*", "", taxid)
  sampleid <- sub("_taxid.*", "", focal_filename)
  file_info <- file.info(focal_dir_filename)
  if(file_info$size > 0) {
    taxid_col <- rep(taxid, length(focal_file[,1]))
    sampleid_col <- rep(sampleid, length(focal_file[,1]))
    focal_file <- cbind(sampleid_col, taxid_col, focal_file)
    coverage_depth_dataframe <- rbind(coverage_depth_dataframe, focal_file)
  }else{
    coverage_depth_dataframe <- coverage_depth_dataframe
  }
}

# Remove some of the zero entries in the alignment stats.
# This happens when aligning the queries against the multiple references.
# Some references are short and no reads align with them.
coverage_depth_dataframe <- coverage_depth_dataframe[coverage_depth_dataframe$numreads > 0,]
# List of all unique taxids:
unique_taxid <- unique(coverage_depth_dataframe$taxid_col)
# List of unique sample-taxon pairs:
coverage_depth_dataframe$taxid_sampleid <- paste0(coverage_depth_dataframe$taxid_col, "_", coverage_depth_dataframe$sampleid_col)
unique_sample_taxid <- unique(coverage_depth_dataframe$taxid_sampleid)

# Calculate the number of queries that aligned with reference per _coverage_stats.txt file (aka sample-taxon pair):
sampletaxid_query_counts <- coverage_depth_dataframe %>%
  group_by(taxid_sampleid) %>%
  dplyr::summarize(sample_total_queries = sum(numreads))
rownames(sampletaxid_query_counts) <- sampletaxid_query_counts$taxid_sampleid
sampletaxid_query_counts <- as.data.frame(sampletaxid_query_counts)

# Calculate the length of references per _coverage_stats.txt file (aka each sample-taxon pairing):
sampletaxid_ref_lengths <- coverage_depth_dataframe %>%
  group_by(taxid_sampleid) %>%
  dplyr::summarize(reference_total_length = sum(endpos))
rownames(sampletaxid_ref_lengths) <- sampletaxid_ref_lengths$taxid_sampleid
sampletaxid_ref_lengths <- as.data.frame(sampletaxid_ref_lengths)

# For loop that goes through the entries from each _coverage_stats.txt file (aka sample-taxon pair)
# And divides the length of each reference by the total length of non-zero references
# This will then be used to scale the addition of the coverage, depth and number of queries stats
for(j in 1:length(unique_sample_taxid)){
  #j <- 3
  focal_entry <- unique_sample_taxid[j]
  df_subset <- coverage_depth_dataframe[coverage_depth_dataframe$taxid_sampleid == focal_entry,] 
  focal_ref_len <- sampletaxid_ref_lengths[focal_entry, 2]
  for(k in 1:length(df_subset[,1])){
    df_subset$ref_prop[k] <- (df_subset$endpos[k]/focal_ref_len)
    df_subset$scaled_coverage[k] <- (df_subset$coverage[k])*df_subset$ref_prop[k]
    df_subset$scaled_depth[k] <- (df_subset$meandepth[k])*df_subset$ref_prop[k]
    df_subset$scaled_queryno[k] <- (df_subset$numreads[k])*df_subset$ref_prop[k]
  }
  subset_summary <- df_subset %>%
    group_by(taxid_sampleid) %>%
    dplyr::summarize(depth_scaled_by_ref = sum(df_subset$scaled_depth),
                     coverage_scaled_by_ref = sum(df_subset$scaled_coverage),
                     queryno_scaled_by_ref = sum(df_subset$scaled_queryno))
  subset_summary$taxid <- regmatches(subset_summary$taxid_sampleid, regexec("_", subset_summary$taxid_sampleid), invert = T)[[1]][1]
  subset_summary$sampleid <- regmatches(subset_summary$taxid_sampleid, regexec("_", subset_summary$taxid_sampleid), invert = T)[[1]][2]
  coverage_depth_dataframe_queries_reduced <- rbind(coverage_depth_dataframe_queries_reduced, subset_summary)
}
taxid_list <- coverage_depth_dataframe_queries_reduced$taxid

# Generates a 2D plot of the depth - coverage relationship for queries that aligned with a reference genome:
QUERY_depth_coverage_relationship <- ggplot(data = coverage_depth_dataframe_queries_reduced, aes(x = coverage_scaled_by_ref, y = depth_scaled_by_ref)) +
  theme_classic() +
  geom_point(size = 2, alpha = 0.5, color = 'black') +
  xlab("Percentage coverage (queries)") +
  ylab("Percentage depth (queries)") +
  #xlim(c(0,25)) +
  #ylim(c(0,2)) +
  theme(axis.text=element_text(size=12))
plot(QUERY_depth_coverage_relationship)
##ggsave(paste0(base_dir, "QUERIES_depth_coverage.png"), width = 7, height = 7)
```

################################################################
Section B: QUERY ALIGNMENT TO REFERENCE WITH BWA VS MTSV ALIGNMENT
################################################################

Takes each of the _query_length.txt files (from MTSv extraction of queries assigned to a particular taxon; PRE alignment), and essentially adds these numbers as a column to the coverage_depth_dataframe_queries_reduced df (this contains the depth & coverage stats POST alignment). That way it is possible to compare the number of queries that were aligned using the bwa and the number of queries that MTSv identified. The number identified by bwa should be less than or equal to the number identified by MTSv.

```{r, message=FALSE}
# QUERIES
query_length_files <- list.files(base_dir, pattern = "_query_lengths.txt$")
mtsv_query_numbers <- data.frame(taxid = numeric(), sample_id = numeric(), mtsv_query_count = numeric())

for(t in 1:length(query_length_files)){
  focal_query_filename <- query_length_files[t]
  taxid <- sub(".*taxid_", "", focal_query_filename)
  taxid <- sub("_query_lengths.txt.*", "", taxid)
  taxid <- as.numeric(taxid)
  focal_query_dir <- paste0(base_dir, focal_query_filename)
  focal_query_file <- read_tsv(focal_query_dir, col_names = FALSE)
  focal_query_num <- as.numeric(focal_query_file$X1)
  sample_id <- sub("_taxid_.*", "", focal_query_filename)
  focal_entry <- c(taxid, sample_id, focal_query_num)
  mtsv_query_numbers <- rbind(mtsv_query_numbers, focal_entry)
  colnames(mtsv_query_numbers) <- c("taxid", "sample_id", "mtsv_query_count")
}

mtsv_query_numbers$taxid_sampleid <- paste0(mtsv_query_numbers$taxid, "_", mtsv_query_numbers$sample_id)
mtsv_query_numbers$mtsv_query_count <- as.numeric(mtsv_query_numbers$mtsv_query_count)
mtsv_query_number_persample <- mtsv_query_numbers %>%
  group_by(sample_id) %>%
  dplyr::summarise(sample_total = sum(mtsv_query_count))

mtsv_query_numbers$sample_total_n <- mtsv_query_number_persample$sample_total[match(mtsv_query_numbers$sample_id, mtsv_query_number_persample$sample_id)]
mtsv_query_numbers$mtsv_read_prop <- mtsv_query_numbers$mtsv_query_count/mtsv_query_numbers$sample_total_n

coverage_depth_dataframe_queries_reduced$queryno_scaled_by_ref <- as.numeric(coverage_depth_dataframe_queries_reduced$queryno_scaled_by_ref)

coverage_depth_dataframe_queries_reduced$mtsv_queryno <- mtsv_query_numbers$mtsv_query_count[match(coverage_depth_dataframe_queries_reduced$taxid_sampleid, mtsv_query_numbers$taxid_sampleid)]

query_plot <- ggplot(data = coverage_depth_dataframe_queries_reduced, aes(x = mtsv_queryno, y = queryno_scaled_by_ref, color = taxid)) +
  theme_classic() +
  geom_point(size = 2) +
  #xlim(0,10000) +
  #ylim(0,10000) +
  geom_abline(slope = 1, intercept = 0) +
  geom_abline(slope = 0, intercept = 0) +
  xlab("No queries per taxon (identified by MTSv)") +
  ylab("No queries per taxon (that aligned with reference\ngenome using bwa)") +
  labs(colour="NCBI taxonomic ID:")

leg <- cowplot::get_legend(query_plot)
leg <- ggplotify::as.ggplot(leg)
query_plot <- query_plot + theme(legend.position = "none")
g <- arrangeGrob(query_plot, leg, nrow= 1) 
plot(g)
#ggsave(paste0(base_dir, "QUERIES_mtsv_vs_aligned.png"), g, width = 10, height = 7)


```


################################################################
#### Section C: DETECTING AGE-RELATED DNA DAMAGE IN QUERIES ####
################################################################

Look for aDNA damage patterns in the queries.
Unlikely to see much as the queries are cut-up reads and do not have the true read termini.
However, if the signal is string, it may be detectable.

```{r}

# Taxa-samples to remove based on the coverage and depth stats:
taxasamples_to_remove <- c()
#taxasamples_to_remove <- c("225326_NHCS-GLCA821AncientBison1-xx-xx-xxx-2019-024-FW_S3.a4b.Gs.lqt.minlen36") # Real one to remove
# List out all the aDNA damage dirs (for queries and reads):
DNA_damage_dirs <- list.dirs(path = base_dir, full.names = F, recursive = F)
# Separate out the query directories and read directories that containt stats on aDNA damage:
read_dir_list_yn <- str_detect(DNA_damage_dirs, "reads_aligned")
DNA_damage_dirs_reads <- DNA_damage_dirs[read_dir_list_yn == T]
DNA_damage_dirs_query <- DNA_damage_dirs[read_dir_list_yn == F]
# Start the list of cumulative 5pCT and 3pAG damage
damage_5p_CT_query <- read.table(file = paste0(base_dir, DNA_damage_dirs_query[1], "/5pCtoT_freq.txt"), sep = '\t', header = TRUE)
damage_3p_AG_query <- read.table(file = paste0(base_dir, DNA_damage_dirs_query[1], "/3pGtoA_freq.txt"), sep = '\t', header = TRUE)
colnames(damage_5p_CT_query) <- c("pos" , "5pC>T")
colnames(damage_3p_AG_query) <- c("pos" , "3pG>A")
damage_5p_CT_query$`5pC>T` <- 0
damage_3p_AG_query$`3pG>A` <- 0
counter_5p <- 0
counter_3p <- 0
ratio_record_query <- data.frame(taxid_sampleid = character(), ratio = numeric())
all_5p_CT_query <- damage_5p_CT_query$pos
all_3p_AG_query <- damage_3p_AG_query$pos
reads_per_taxid <- data.frame(taxa_id = numeric(), 
                              damage_5pCT_ratio = numeric(), 
                              average_5pCT_ratio_5bp = numeric(),
                              damage_3pAG_ratio = numeric(),
                              average3pAG_ratio = numeric())

for(j in 1:length(DNA_damage_dirs_query)){
  focal_dir <- DNA_damage_dirs_query[j]
  taxid <- sub(".*taxid_", "", focal_dir)
  taxid <- sub("_aligned.*", "", taxid) 
  sampleid <- sub("_taxid_.*", "", focal_dir)
  sampleid <- sub(".*results_", "", sampleid)
  taxid_sampleid <- paste0(taxid, "_", sampleid)
  if(taxid_sampleid %in% taxasamples_to_remove == F) {
    focal_5p_file <- read.table(file = paste0(base_dir, focal_dir, "/5pCtoT_freq.txt"), sep = '\t', header = TRUE)
    focal_3p_file <- read.table(file = paste0(base_dir, focal_dir, "/3pGtoA_freq.txt"), sep = '\t', header = TRUE)
    damage_5p_CT_query$`5pC>T` <- damage_5p_CT_query$`5pC>T` + focal_5p_file$X5pC.T
    damage_3p_AG_query$`3pG>A` <- damage_3p_AG_query$`3pG>A` + focal_3p_file$X3pG.A
    damage_5pCT_ratio <- (sum(damage_5p_CT_query[1:5,"5pC>T"])) /(sum(damage_5p_CT_query[6:10,"5pC>T"]))
    damage_5pCT_ratio_5bp <- mean(damage_5p_CT_query[1:5,"5pC>T"])
    damage_3pAG_ratio <- (sum(damage_3p_AG_query[1:5,"3pG>A"])) /(sum(damage_3p_AG_query[6:10,"3pG>A"]))
    damage_3pAG_ratio_5bp <- mean(damage_3p_AG_query[1:5,"3pG>A"])
    x <- as.data.frame(cbind(taxid_sampleid, damage_5pCT_ratio, damage_5pCT_ratio_5bp, damage_3pAG_ratio, damage_3pAG_ratio_5bp))
    ratio_record_query <- rbind(ratio_record_query, x)
    y <- focal_5p_file$X5pC.T
    y <- as.data.frame(y)
    colnames(y) <- taxid_sampleid
    all_5p_CT_query <- cbind(all_5p_CT_query, y)
    z <- focal_3p_file$X3pG.A
    z <- as.data.frame(z)
    colnames(z) <- taxid_sampleid
    all_3p_AG_query <- cbind(all_3p_AG_query, z)
  }else{
    ratio_record_query <- ratio_record_query
    all_5p_CT_query <- all_5p_CT_query
    all_3p_AG_query <- all_3p_AG_query
  }
}

coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio <- ratio_record_query$damage_5pCT_ratio[match(coverage_depth_dataframe_queries_reduced$taxid_sampleid, ratio_record_query$taxid_sampleid)]
coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio <- ratio_record_query$damage_3pAG_ratio[match(coverage_depth_dataframe_queries_reduced$taxid_sampleid, ratio_record_query$taxid_sampleid)]

damage_5p_CT_query$`5pC>T` <- damage_5p_CT_query$`5pC>T` / length(DNA_damage_dirs_query)
damage_3p_AG_query$`3pG>A` <- damage_3p_AG_query$`3pG>A` / length(DNA_damage_dirs_query)
total_damage_query <- cbind(damage_5p_CT_query, damage_3p_AG_query$`3pG>A`)
colnames(total_damage_query) <- c("pos" , "5pC>T", "3pG>A")

damage_plot_5p3p_query <- ggplot(total_damage_query) +
  geom_line(aes(x = pos, y = `5pC>T`, color = 'C -> T at 5` terminal'), size = 2) +
  #geom_smooth(aes(x = pos, y = `5pC>T`, color = 'C -> T at 5` terminal'), size = 2, alpha = 0.5) +
  geom_line(aes(x = pos, y = `3pG>A`, color = 'A -> G at 5` terminal'), size = 2) + 
  #geom_smooth(aes(x = pos, y = `3pG>A`, color = 'A -> G at 3` terminal'), size = 2, alpha = 0.5) + 
  theme_classic() +
  xlab("Distance from read terminal (bp)") +
  ylab("Rate of nucleotide transition per bp") +
  scale_color_manual(name = "Nucleotide transition rate:",
                     values = c("darkblue", "darkred"))
plot(damage_plot_5p3p_query)
#ggsave(paste0(base_dir, "QUERIES_nuc_transition_rate_mean.png"), width = 10, height = 7)

coverage_depth_dataframe_queries_reduced$coverage_scaled_by_ref <- as.numeric(coverage_depth_dataframe_queries_reduced$coverage_scaled_by_ref)
coverage_depth_dataframe_queries_reduced$depth_scaled_by_ref <- as.numeric(coverage_depth_dataframe_queries_reduced$depth_scaled_by_ref)
coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio <- as.numeric(coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio)
coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio <- as.numeric(coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio)


{
coverage_CT_plot_query <- ggplot(coverage_depth_dataframe_queries_reduced) +
  theme_classic() +
  geom_point(aes(x=coverage_scaled_by_ref, y=damage_5pCT_ratio, color = taxid)) +
  xlab("Average coverage across reference genome") +
  ylab("Rate of C -> T in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_queries_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1)

leg <- cowplot::get_legend(coverage_CT_plot_query)
leg <- ggplotify::as.ggplot(leg)
coverage_CT_plot_query <- coverage_CT_plot_query + theme(legend.position = "none")

coverage_AG_plot_query <- ggplot(coverage_depth_dataframe_queries_reduced) +
  theme_classic() +
  geom_point(aes(x=coverage_scaled_by_ref, y=damage_3pAG_ratio, color = taxid)) +
  xlab("Average coverage across reference genome") +
  ylab("Rate of A -> G in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_queries_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1) +
  scale_x_reverse() +
  scale_y_continuous(position = "right")
coverage_AG_plot_query <- coverage_AG_plot_query + theme(legend.position = "none")

max_coverage <- max(max(coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio), 
                    max(coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio))

coverage_CT_plot_query <- coverage_CT_plot_query + ylim(c(0,max_coverage))
coverage_AG_plot_query <- coverage_AG_plot_query + ylim(c(0,max_coverage))

g <- arrangeGrob(coverage_CT_plot_query, coverage_AG_plot_query, leg, nrow= 1) 
plot(g)
#ggsave(file= paste0(base_dir, "QUERIES_rateCT&AG_vs_coverage.png"), g, width = 14, height = 7)


}


{
depth_CT_plot_query <- ggplot(coverage_depth_dataframe_queries_reduced) +
  theme_classic() +
  geom_point(aes(x=depth_scaled_by_ref, y=damage_5pCT_ratio, color = taxid)) +
  xlab("Average depth across reference genome") +
  ylab("Rate of C -> T in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_queries_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1)

leg <- cowplot::get_legend(depth_CT_plot_query)
leg <- ggplotify::as.ggplot(leg)
depth_CT_plot_query <- depth_CT_plot_query + theme(legend.position = "none")


depth_AG_plot_query <- ggplot(coverage_depth_dataframe_queries_reduced) +
  theme_classic() +
  geom_point(aes(x=depth_scaled_by_ref, y=damage_3pAG_ratio, color = taxid)) +
  xlab("Average depth across reference genome") +
  ylab("Rate of A -> G in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  ylim(c(0, NA)) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_queries_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1) +
  scale_x_reverse() +
  scale_y_continuous(position = "right")

depth_AG_plot_query <- depth_AG_plot_query + theme(legend.position = "none")
max_depth <- max(max(coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio), 
                    max(coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio))
depth_CT_plot_query <- depth_CT_plot_query + ylim(c(0,max_depth))
depth_AG_plot_query <- depth_AG_plot_query + ylim(c(0,max_depth))

g <- arrangeGrob(depth_CT_plot_query, depth_AG_plot_query, leg, nrow= 1) 
plot(g)
#ggsave(file= paste0(base_dir, "QUERIES_rateCT&AG_vs_depth.png"), g, width = 14, height = 7)

}

```



```{r}

melted_ratio_record <- melt(ratio_record_query, id = "taxid_sampleid")
melted_ratio_record_5p <- melted_ratio_record[melted_ratio_record$variable == "damage_5pCT_ratio",]
melted_ratio_record_3p <- melted_ratio_record[melted_ratio_record$variable == "damage_3pAG_ratio",]
melted_ratio_record_both <- rbind(melted_ratio_record_3p, melted_ratio_record_5p)
melted_ratio_record_both <- as.data.frame(melted_ratio_record_both)
melted_ratio_record_both$variable <- as.character(melted_ratio_record_both$variable)
melted_ratio_record_both$variable[melted_ratio_record_both$variable == "damage_5pCT_ratio"] <- "5'"
melted_ratio_record_both$variable[melted_ratio_record_both$variable == "damage_3pAG_ratio"] <- "3'"
colnames(melted_ratio_record_both) <- c("taxid_sampleid", "sense", "ratio")
melted_ratio_record_both$ratio <- as.numeric(melted_ratio_record_both$ratio)

ratio_record_plot <- ggplot(melted_ratio_record_both, aes(x = sense, y = ratio, group = sense, color = sense, fill = sense)) + 
  theme_classic() +
  ylab("Ratio of C->T and A->G transitions\nin terminal bps 0-5 over 6-10") +
  xlab ("Read terminal identity") + 
  theme(legend.position = "none") +
  ylim(c(0,NA)) + 
  geom_abline(slope = 0, intercept = 1)
ratio_record_plot <- ratio_record_plot + geom_violin(trim = FALSE) 
plot(ratio_record_plot)
#ggsave(paste0(base_dir, "QUERIES_ratio_CT_AG.png"), width = 5, height = 7)

all_5p_CT_query <- cbind(all_5p_CT_query, damage_5p_CT_query$`5pC>T`)
all_3p_AG_query <- cbind(all_3p_AG_query, damage_3p_AG_query$`3pG>A`)
names(all_5p_CT_query)[names(all_5p_CT_query) == 'damage_5p_CT_query$`5pC>T`'] <- 'Total 5pC>T'
names(all_3p_AG_query)[names(all_3p_AG_query) == 'damage_3p_AG_query$`3pG>A`'] <- 'Total 3pG>A'

# Melts the dataframes, to make plotting the data easier
melted_all_CT_query <- melt(all_5p_CT_query, id = "all_5p_CT_query")
melted_all_AG_query <- melt(all_3p_AG_query, id = "all_3p_AG_query")

# Assign color to each of the sample-taxa that went through DNA damage detection:
set.seed(4)
n_col_df <- melted_all_CT_query %>%
  group_by(variable) %>%
  dplyr::summarise(n = n())
n_col_df <- as.data.frame(n_col_df)
n_col <- length(n_col_df[,1])  

col_vector2 <- distinctColorPalette(k = n_col, altCol = FALSE, runTsne = FALSE)
n_col_df$color_col <- col_vector2
melted_all_CT_query$col_column <- n_col_df$color_col[match(melted_all_CT_query$variable, n_col_df$variable)]

set.seed(4)
n_col_df <- melted_all_AG_query %>%
  group_by(variable) %>%
  dplyr::summarise(n = n())
n_col_df <- as.data.frame(n_col_df)
n_col <- length(n_col_df[,1])  

col_vector3 <- distinctColorPalette(k = n_col, altCol = FALSE, runTsne = FALSE)
n_col_df$color_col <- col_vector3
melted_all_AG_query$col_column <- n_col_df$color_col[match(melted_all_AG_query$variable, n_col_df$variable)]

melted_all_CT_query <- as.data.frame(melted_all_CT_query)
names(melted_all_CT_query)[names(melted_all_CT_query) == 'variable'] <- 'taxid_sampleid'
names(melted_all_CT_query)[names(melted_all_CT_query) == 'all_5p_CT_query'] <- 'read_pos'
melted_all_CT_query$col_column[melted_all_CT_query$taxid == 'Total 5pC>T'] <- '#000000'
melted_all_CT_query$line_thickness <- 0.1
melted_all_CT_query$line_thickness[melted_all_CT_query$taxid == 'Total 5pC>T'] <- 0.5
melted_all_CT_query$col_column <- c(melted_all_CT_query$col_column)

melted_all_AG_query <- as.data.frame(melted_all_AG_query)
names(melted_all_AG_query)[names(melted_all_AG_query) == 'variable'] <- 'taxid_sampleid'
names(melted_all_AG_query)[names(melted_all_AG_query) == 'all_3p_AG_query'] <- 'read_pos'
melted_all_AG_query$col_column[melted_all_AG_query$taxid == 'Total 3pG>A'] <- '#000000'
melted_all_AG_query$line_thickness <- 0.1
melted_all_AG_query$line_thickness[melted_all_AG_query$taxid == 'Total 3pG>A'] <- 0.5

melted_all_CT_query$col_column[melted_all_CT_query$col_column != "#000000"] <- "skyblue"
melted_all_AG_query$col_column[melted_all_AG_query$col_column != "#000000"] <- "skyblue"


{
melted_all_CT_plot_query <- ggplot(melted_all_CT_query, aes(x = read_pos, y = value)) +
  theme_classic() +
  geom_line(aes(size = line_thickness,  group = taxid_sampleid,  color = I(col_column))) +
  xlab("Distance from 5' end of read (bp)") +
  ylab("Rate of C -> T transition per bp") +
  ylim(0, 0.1) +
  scale_size_continuous(name = "Rate of nucleotide\ntranistions:", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) 
  
leg <- cowplot::get_legend(melted_all_CT_plot_query)
leg <- ggplotify::as.ggplot(leg)
melted_all_CT_plot_query <- melted_all_CT_plot_query + theme(legend.position = "none")

melted_all_AG_plot_query <- ggplot(melted_all_AG_query, aes(x = read_pos, y = value)) +
  theme_classic() +
  geom_line(aes(size = line_thickness,  group = taxid_sampleid,  color = I(col_column))) + 
  xlab("Distance from 3' end of read (bp)") +
  ylab("Rate of A -> G transition per bp") +
  scale_size_continuous(name = "Rate of nucleotide\ntranistions:", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) +
  scale_x_reverse() +
  scale_y_continuous(limits = c(0,0.1), position = "right")

melted_all_AG_plot_query <- melted_all_AG_plot_query + theme(legend.position = "none")

g <- arrangeGrob(melted_all_CT_plot_query, melted_all_AG_plot_query, leg, nrow=1) 
plot(g)
#ggsave(file= paste0(base_dir, "QUERIES_CT_AG_transition_perTaxon.png"), g, width = 14, height = 7)
 
}


```



```{r}

##########################################
# Scaling the number C->T rates for each taxon-sample pair by the relative proportion of queries assigned to that pairing versus the total for all pairings:

all_mtsv_query_no <- all_5p_CT_query[0:1, 2:length(all_5p_CT_query[1,])]
all_mtsv_query_no[1,] <- 0
for(j in 1:length(DNA_damage_dirs_query)){
  #j <- 3
  focal_dir <- DNA_damage_dirs_query[j]
  taxid <- sub(".*taxid_", "", focal_dir)
  taxid <- sub("_aligned.*", "", taxid) 
  sampleid <- sub("_taxid_.*", "", focal_dir)
  sampleid <- sub(".*results_", "", sampleid)
  taxid_sampleid <- paste0(taxid, "_", sampleid)
  #print(taxid_sampleid)
  query_len_file <- paste0(sampleid, "_taxid_", taxid, "_query_lengths.txt")
  #print(query_len_file)
  if(taxid_sampleid %in% taxasamples_to_remove == F) {
    no_mtsv_queries <- read.table(file = paste0(base_dir, query_len_file), sep = '\t', header = F)
    all_mtsv_query_no[1, taxid_sampleid] <- no_mtsv_queries$V1
  }else{
    all_mtsv_query_no <- all_mtsv_query_no
  }
}
#all_mtsv_query_no <- all_mtsv_query_no[, colSums(all_mtsv_query_no != 0) > 0]
total_mtsv_queries <- sum(all_mtsv_query_no)
# Creating a duplicate of the all_5p_CT_query data frame, which will be scaled based on the proportion of queries in each sample-taxon pair relative to the total number of queries across all sample-taxon pairs
all_5p_CT_query_2 <- all_5p_CT_query
damage_5p_CT_query_2 <- damage_5p_CT_query
damage_5p_CT_query_2$`5pC>T` <- 0
mtsv_query_prop <- all_mtsv_query_no/total_mtsv_queries
common_taxa <- intersect(colnames(all_5p_CT_query_2), colnames(mtsv_query_prop))
mtsv_query_prop <- mtsv_query_prop[, colnames(mtsv_query_prop) == common_taxa]

for(i in 2:(length(all_5p_CT_query[1,])-1)){
  focal_taxa_sample <- colnames(all_5p_CT_query[i])
  prop_mtsv <- mtsv_query_prop[1,colnames(mtsv_query_prop) == focal_taxa_sample]
  all_5p_CT_query_2[,focal_taxa_sample] <- all_5p_CT_query_2[,focal_taxa_sample] * prop_mtsv
}


all_5p_CT_query_2$`Total 5pC>T` <- rowSums(all_5p_CT_query_2[,2:(length(all_5p_CT_query_2)-1)])
damage_5p_CT_query_2$`5pC>T` <- all_5p_CT_query_2$`Total 5pC>T`

all_3p_AG_query_2 <- all_3p_AG_query
damage_3p_AG_query_2 <- damage_3p_AG_query
damage_3p_AG_query_2$`3pG>A` <- 0
mtsv_query_prop <- all_mtsv_query_no/total_mtsv_queries
common_taxa <- intersect(colnames(all_3p_AG_query_2), colnames(mtsv_query_prop))
mtsv_query_prop <- mtsv_query_prop[, colnames(mtsv_query_prop) == common_taxa]

for(i in 2:(length(all_3p_AG_query_2[1,])-1)){
  focal_taxa_sample <- colnames(all_3p_AG_query_2[i])
  prop_mtsv <- mtsv_query_prop[1,colnames(mtsv_query_prop) == focal_taxa_sample]
  all_3p_AG_query_2[,focal_taxa_sample] <- all_3p_AG_query_2[,focal_taxa_sample] * prop_mtsv
}

all_3p_AG_query_2$`Total 3pG>A` <- rowSums(all_3p_AG_query_2[,2:(length(all_3p_AG_query_2)-1)])
damage_3p_AG_query_2$`3pG>A` <- all_3p_AG_query_2$`Total 3pG>A`


CT_plot_scaled_query_no <- ggplot(damage_5p_CT_query_2,  aes(x = pos, y = `5pC>T`)) +
  theme_classic() +
  geom_smooth(color = 'dark red', size = 1) +
  #geom_smooth(color = 'black', size = 0.5, linetype = 'dashed') + 
  #geom_line(color = 'dark red', size = 1) + 
  xlab("Distance from 5' end of read (bp)") +
  ylab("Rate of C -> T transition per bp, scaled by read number") +
  scale_size_continuous(name = "Rate of C -> T transition\nper base at 5'", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) +
  ylim(0,0.1)
#plot(CT_plot_scaled_query_no)


AG_plot_scaled_query_no <- ggplot(damage_3p_AG_query_2,  aes(x = pos, y = `3pG>A`)) +
  theme_classic() +
  geom_smooth(color = 'blue', size = 1) +
  #geom_smooth(color = 'black', size = 0.5, linetype = 'dashed') + 
  #geom_line(color = 'blue', size = 1) + 
  xlab("Distance from 3' end of read (bp)") +
  ylab("Rate of A -> G transition per bp, scaled by read number") +
  scale_size_continuous(name = "Rate of A -> G transition\nper base at 3'", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) +
  scale_x_reverse() +
  scale_y_continuous(limits = c(0,0.1), position = "right") 
#plot(AG_plot_scaled_query_no)

#grid.arrange(CT_plot_scaled_query_no, AG_plot_scaled_query_no, ncol = 2)
#ggsave(paste0(base_dir, "QUERIES_CT_AG_transition_perQueryNo.png"), width = 14, height = 7)
g <- arrangeGrob(CT_plot_scaled_query_no, AG_plot_scaled_query_no, nrow=1) 
plot(g)
#ggsave(file=paste0(base_dir, "QUERIES_CT_AG_transition_perQueryNo.png"), g, width = 14, height = 7)


saveRDS(damage_5p_CT_query_2, file = paste0(base_dir, "query_damage_5p_CT_2.rds"))
saveRDS(damage_3p_AG_query_2, file = paste0(base_dir, "query_damage_3p_AG_2.rds"))
saveRDS(damage_5p_CT_query, file = paste0(base_dir, "query_damage_5p_CT.rds"))
saveRDS(damage_3p_AG_query, file = paste0(base_dir, "query_damage_3p_AG.rds"))
```




```{r}
{
  
max_ratio <- max(max(coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio), max(coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio))
  
QUERY_depth_coverage_5p_relationship <- ggplot(data = coverage_depth_dataframe_queries_reduced, aes(x = coverage_scaled_by_ref, y = depth_scaled_by_ref)) +
  theme_classic() +
  geom_point(aes(color = coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio), size = 2, alpha = 1) +
  xlab("Percentage coverage (queries)") +
  ylab("Percentage depth (queries)") +
  xlim(c(0,25)) +
  ylim(c(0,1)) +
  theme(axis.text=element_text(size=12)) +
  scale_color_gradientn(colours = c("darkred","white","royalblue", "royalblue4","darkblue", "black"), 
                        values = c(0,  max(damage_3pAG_ratio)/(max_ratio*(3/5)), 1),
                        guide = "colorbar", limits=c(0, max_ratio),
                        name = "Cumulative nucleotide transition rate\nin terminal bases 1-5 over 6-10")
  
leg <- cowplot::get_legend(QUERY_depth_coverage_5p_relationship)
leg <- ggplotify::as.ggplot(leg)
QUERY_depth_coverage_5p_relationship <- QUERY_depth_coverage_5p_relationship + theme(legend.position = "none")

QUERY_depth_coverage_3p_relationship <- ggplot(data = coverage_depth_dataframe_queries_reduced, aes(x = coverage_scaled_by_ref, y = depth_scaled_by_ref)) +
  theme_classic() +
  geom_point(aes(color = coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio), size = 2, alpha = 1) +
  xlab("Percentage coverage (queries)") +
  ylab("Percentage depth (queries)") +
  xlim(c(0,25)) +
  ylim(c(0,1)) +
  theme(axis.text=element_text(size=12)) +
  scale_color_gradientn(colours = c("darkred","white","royalblue", "royalblue4","darkblue", "black"), 
                        values = c(0,  max(damage_3pAG_ratio)/(max_ratio*(3/5)), 1),
                        guide = "colorbar", limits=c(0, max_ratio),
                        name = "Cumulative nucleotide transition rate\nin terminal bases 1-5 over 6-10")
QUERY_depth_coverage_3p_relationship <- QUERY_depth_coverage_3p_relationship + 
                                        theme(legend.position = "none",
                                              axis.title.y=element_blank())

g <- arrangeGrob(QUERY_depth_coverage_5p_relationship, QUERY_depth_coverage_3p_relationship, leg, nrow=1) 
plot(g)
#ggsave(file = paste0(base_dir, "QUERIES_depth_coverage_5p3p_relationship.png"), g, width = 14, height = 7)

}
```



################################################################
#### Section D: READS DATAFRAME - COVERAGE AND DEPTH STATS #####
################################################################

The read depth and coverage stats were generated by:
  - Extracting queries from MTSv.
  - Grepping these back against the reads (linearized FASTQ files)
  - The matching linear FASTQ files that matched were reformatted.
  - And aligned against a reference genome (where available) using: BWA aln.
  - SAMTOOLS was used to calculate the depth and coverage stats in each _reads_coverage_stats.txt file.

```{r, message=FALSE}
# Empty df that will contain stats from each of the separate stats.txt files:
coverage_depth_dataframe <- data.frame(sampleid = character(),
                                       taxid = integer(), 
                                       rname = character(),
                                       startpos = numeric(),
                                       endpos = numeric(),
                                       numreads = numeric(),
                                       covbases = numeric(),
                                       coverage = numeric(),
                                       meandepth = numeric(),
                                       meanbaseq = numeric(),
                                       meanmapq = numeric())   
# For loop below goes through each of the _reads_coverage_stats.txt files and adds hem to the df above.
# In many cases there are multiple sequences per reference, resulting in multiple lines per _reads_coverage_stats.txt file
# Empty df below: will contain the summary statistics from the _reads_coverage_stats.txt files, 
# i.e. it will have a single line for each set of taxon reads that were extracted by MTSv 
coverage_depth_dataframe_reads_reduced <- data.frame(taxid = numeric(), 
                                                     sampleid = character(),
                                                     scaled_depth = numeric(),
                                                     scaled_coverage = numeric(),
                                                     readno_scaled_by_ref = numeric(),
                                                     taxid_sampleid = character())
for(i in 1:length(readfile_list)){
  #i <- 1
  focal_filename <- readfile_list[i]
  focal_dir_filename <- paste0(base_dir, focal_filename)
  focal_file <- read_tsv(focal_dir_filename, col_names = TRUE)
  focal_file <- as.data.frame(focal_file)
  taxid <- sub(".*taxid_", "", focal_filename)
  taxid <- sub("_reads_coverage_stats.txt.*", "", taxid)
  sampleid <- sub("_taxid.*", "", focal_filename)
  file_info <- file.info(focal_dir_filename)
  if(file_info$size > 0) {
    taxid_col <- rep(taxid, length(focal_file[,1]))
    sampleid_col <- rep(sampleid, length(focal_file[,1]))
    focal_file <- cbind(sampleid_col, taxid_col, focal_file)
    coverage_depth_dataframe <- rbind(coverage_depth_dataframe, focal_file)
  }else{
    coverage_depth_dataframe <- coverage_depth_dataframe
  }
}
# Remove some of the zero entries in the alignment stats.
# This happens when aligning the reads against the multiple references.
# Some references are short and no reads align with them.
coverage_depth_dataframe <- coverage_depth_dataframe[coverage_depth_dataframe$numreads > 0,]
# List of all unique taxids:
unique_taxid <- unique(coverage_depth_dataframe$taxid_col)
# List of unique sample-taxon pairs:
coverage_depth_dataframe$taxid_sampleid <- paste0(coverage_depth_dataframe$taxid_col, "_", coverage_depth_dataframe$sampleid_col)
unique_sample_taxid <- unique(coverage_depth_dataframe$taxid_sampleid)

# Calculate the number of reads that aligned with reference per _reads_coverage_stats.txt file (aka sample-taxon pair):
sampletaxid_read_counts <- coverage_depth_dataframe %>%
  group_by(taxid_sampleid) %>%
  dplyr::summarize(sample_total_reads = sum(numreads))
rownames(sampletaxid_read_counts) <- sampletaxid_read_counts$taxid_sampleid
sampletaxid_read_counts <- as.data.frame(sampletaxid_read_counts)

# Calculate the length of references per _reads_coverage_stats.txt file (aka each sample-taxon pairing):
sampletaxid_ref_lengths <- coverage_depth_dataframe %>%
  group_by(taxid_sampleid) %>%
  dplyr::summarize(reference_total_length = sum(endpos))
rownames(sampletaxid_ref_lengths) <- sampletaxid_ref_lengths$taxid_sampleid
sampletaxid_ref_lengths <- as.data.frame(sampletaxid_ref_lengths)

# For loop that goes through the entries from each _reads_coverage_stats.txt file (aka sample-taxon pair)
# And divides the length of each reference by the total length of non-zero references
# This will then be used to scale the addition of the coverage, depth and number of reads stats
for(j in 1:length(unique_sample_taxid)){
  #j <- 3
  focal_entry <- unique_sample_taxid[j]
  df_subset <- coverage_depth_dataframe[coverage_depth_dataframe$taxid_sampleid == focal_entry,] 
  focal_ref_len <- sampletaxid_ref_lengths[focal_entry, 2]
  for(k in 1:length(df_subset[,1])){
    df_subset$ref_prop[k] <- (df_subset$endpos[k]/focal_ref_len)
    df_subset$scaled_coverage[k] <- (df_subset$coverage[k])*df_subset$ref_prop[k]
    df_subset$scaled_depth[k] <- (df_subset$meandepth[k])*df_subset$ref_prop[k]
    df_subset$scaled_readno[k] <- (df_subset$numreads[k])*df_subset$ref_prop[k]
  }
  subset_summary <- df_subset %>%
    group_by(taxid_sampleid) %>%
    dplyr::summarize(depth_scaled_by_ref = sum(df_subset$scaled_depth),
                     coverage_scaled_by_ref = sum(df_subset$scaled_coverage),
                     readno_scaled_by_ref = sum(numreads))
  subset_summary$taxid <- regmatches(subset_summary$taxid_sampleid, regexec("_", subset_summary$taxid_sampleid), invert = T)[[1]][1]
  subset_summary$sampleid <- regmatches(subset_summary$taxid_sampleid, regexec("_", subset_summary$taxid_sampleid), invert = T)[[1]][2]
  coverage_depth_dataframe_reads_reduced <- rbind(coverage_depth_dataframe_reads_reduced, subset_summary)
}
taxid_list <- coverage_depth_dataframe_reads_reduced$taxid


# Generates a 2D plot of the depth - coverage relationship for reads that aligned with a reference genome:
READS_depth_coverage_relationship <- ggplot(data = coverage_depth_dataframe_reads_reduced, aes(x = coverage_scaled_by_ref, y = depth_scaled_by_ref)) +
  theme_classic() +
  geom_point(size = 2, alpha = 0.5, color = 'black') +
  xlab("Percentage coverage (queries)") +
  ylab("Percentage depth (queries)") +
  xlim(c(0,25)) +
  ylim(c(0,2)) +
  theme(axis.text=element_text(size=12))

plot(READS_depth_coverage_relationship)
##ggsave(paste0(base_dir, "READS_depth_coverage.png"), width = 10, height = 7)

```

################################################################
Section E: READS ALIGNMENT TO REFERENCE WITH BWA VS MTSV ALIGNMENT
################################################################

Takes each of the _reads_length.txt files (from MTSv extraction + grepped against reads PRE alignment), and essentially adds this as a column to the coverage_depth_dataframe_reads_reduced df (this contains the depth, coverage stats POST alignment). That way it is possible to compare the number of reads that were aligned using the bwa and the number of 'reads' that MTSv identified. However, MTSv does not really identify reads, it identifies queries

```{r, message=FALSE}
# GO THROUGH EACH OF THE QUERY AND READ LENGTH FILES (POST-EXTRACTION ALIGNMENT)  
# AND COMPARE THESE TO THE NUMBER OF FILES USED BY MTSV (EXTRACTED)

# READS:
read_length_files <- list.files(base_dir, pattern = "_read_lengths.txt$")
mtsv_read_numbers <- data.frame(taxid = numeric(), sample_id = numeric(), mtsv_read_count = numeric())

for(t in 1:length(read_length_files)){
  #t <- 1
  focal_read_filename <- read_length_files[t]
  taxid <- sub(".*taxid_", "", focal_read_filename)
  taxid <- sub("_read_lengths.txt.*", "", taxid)
  taxid <- as.numeric(taxid)
  focal_read_dir <- paste0(base_dir, focal_read_filename)
  focal_read_file <- read_tsv(focal_read_dir, col_names = FALSE)
  focal_read_file <- as.data.frame(focal_read_file)
  focal_read_num <- as.numeric(focal_read_file$X1)
  sample_id <- sub("_taxid_.*", "", focal_read_filename)
  focal_entry <- c(taxid, sample_id, focal_read_num)
  mtsv_read_numbers <- rbind(mtsv_read_numbers, focal_entry)
  colnames(mtsv_read_numbers) <- c("taxid", "sample_id", "mtsv_read_count")
}

mtsv_read_numbers$taxid_sampleid <- paste0(mtsv_read_numbers$taxid, "_", mtsv_read_numbers$sample_id)
mtsv_read_numbers$mtsv_read_count <- as.numeric(mtsv_read_numbers$mtsv_read_count)
mtsv_read_number_persample <- mtsv_read_numbers %>%
  group_by(sample_id) %>%
  dplyr::summarise(sample_total = sum(mtsv_read_count))

mtsv_read_numbers$sample_total_n <- mtsv_read_number_persample$sample_total[match(mtsv_read_numbers$sample_id, mtsv_read_number_persample$sample_id)]
mtsv_read_numbers$mtsv_read_prop <- mtsv_read_numbers$mtsv_read_count/mtsv_read_numbers$sample_total_n

coverage_depth_dataframe_reads_reduced$readno_scaled_by_ref <- as.numeric(coverage_depth_dataframe_reads_reduced$readno_scaled_by_ref)


coverage_depth_dataframe_reads_reduced$mtsv_readno <- mtsv_read_numbers$mtsv_read_count[match(coverage_depth_dataframe_reads_reduced$taxid_sampleid, mtsv_read_numbers$taxid_sampleid)]
coverage_depth_dataframe_reads_reduced$mtsv_readprop <- mtsv_read_numbers$mtsv_read_prop[match(coverage_depth_dataframe_reads_reduced$taxid_sampleid, mtsv_read_numbers$taxid_sampleid)]

# This plot shows the relationship between the number of reads identified by MTSv (+ grep) and the reads that aligned with the reference genome using bwa aln
reads_plot <- ggplot(data = coverage_depth_dataframe_reads_reduced, aes(x = mtsv_readno, y = readno_scaled_by_ref, color = taxid)) +
  theme_classic() +
  geom_point(size = 2) +
  xlim(0,10000) +
  ylim(0,10000) +
  geom_abline(slope = 1, intercept = 0) +
  geom_abline(slope = 0, intercept = 0) +
  labs(colour="Taxanomic ID") +
  xlab("No reads per taxon (identified by MTSv)") +
  ylab("No reads per taxon (that aligned with reference\ngenome using bwa)") 
plot(reads_plot)
#ggsave(paste0(base_dir, "READS_mtsv_vs_aligned.png"), width = 10, height = 7)

coverage_depth_dataframe_reads_reduced$queryno_scaled_by_ref <- coverage_depth_dataframe_queries_reduced$queryno_scaled_by_ref[match(coverage_depth_dataframe_reads_reduced$taxid_sampleid, coverage_depth_dataframe_queries_reduced$taxid_sampleid)]
aligned_readsVqueries_plot <- ggplot(data = coverage_depth_dataframe_reads_reduced, aes(x = queryno_scaled_by_ref, y = readno_scaled_by_ref, color = taxid)) +
  theme_classic() +
  geom_point(size = 2) +
  xlim(0,10000) +
  ylim(0,10000) +
  geom_abline(slope = 1, intercept = 0) +
  geom_abline(slope = 0, intercept = 0) +
  labs(colour="Taxanomic ID") +
  xlab("No QUERIES per taxon (that aligned with reference\ngenome using bwa)") +
  ylab("No READS per taxon (that aligned with reference\ngenome using bwa)") 
plot(aligned_readsVqueries_plot)
#ggsave(paste0(base_dir, "READS_aligned_vs_QUERIES_aligned.png"), width = 10, height = 7)


coverage_depth_dataframe_reads_reduced$mtsv_queryno <- coverage_depth_dataframe_queries_reduced$mtsv_queryno[match(coverage_depth_dataframe_reads_reduced$taxid_sampleid, coverage_depth_dataframe_queries_reduced$taxid_sampleid)]
mtsv_readsVSqueries_plot <- ggplot(data = coverage_depth_dataframe_reads_reduced, aes(x = mtsv_queryno, y = mtsv_readno, color = taxid)) +
  theme_classic() +
  geom_point(size = 2) +
  xlim(0,10000) +
  ylim(0,10000) +
  geom_abline(slope = 1, intercept = 0) +
  geom_abline(slope = 0, intercept = 0) +
  labs(colour="Taxanomic ID") +
  xlab("No QUERIES per taxon (identified by MTSv)") +
  ylab("No READS per taxon (identified by MTSv)") 
plot(mtsv_readsVSqueries_plot)
#ggsave(paste0(base_dir, "READS_mtsv_vs_QUERIES_mtsv.png"), width = 10, height = 7)

```


################################################################
##### Section F: DETECTING AGE-RELATED DNA DAMAGE IN READS #####
################################################################

```{r}
# AMALGAMATE DNA DAMAGE:
damage_5p_CT <- read.table(file = paste0(base_dir, DNA_damage_dirs_reads[1], "/5pCtoT_freq.txt"), sep = '\t', header = TRUE)
damage_3p_AG <- read.table(file = paste0(base_dir, DNA_damage_dirs_reads[1], "/3pGtoA_freq.txt"), sep = '\t', header = TRUE)
colnames(damage_5p_CT) <- c("pos" , "5pC>T")
colnames(damage_3p_AG) <- c("pos" , "3pG>A")
damage_5p_CT$`5pC>T` <- 0
damage_3p_AG$`3pG>A` <- 0
counter_5p <- 0
counter_3p <- 0
ratio_record <- data.frame(taxid_sampleid = character(), ratio = numeric())
all_5p_CT <- damage_5p_CT$pos
all_3p_AG <- damage_3p_AG$pos
reads_per_taxid <- data.frame(taxa_id = numeric(), 
                              damage_5pCT_ratio = numeric(), 
                              average_5pCT_ratio_5bp = numeric(),
                              damage_3pAG_ratio = numeric(),
                              average3pAG_ratio = numeric())

for(j in 1:length(DNA_damage_dirs_reads)){

  focal_dir <- DNA_damage_dirs_reads[j]
  taxid <- sub(".*taxid_", "", focal_dir)
  taxid <- sub("_reads_aligned.*", "", taxid) 
  sampleid <- sub("_taxid_.*", "", focal_dir)
  sampleid <- sub(".*results_", "", sampleid)
  taxid_sampleid <- paste0(taxid, "_", sampleid)
  
  focal_5p_file <- read.table(file = paste0(base_dir, focal_dir, "/5pCtoT_freq.txt"), sep = '\t', header = TRUE)
  focal_3p_file <- read.table(file = paste0(base_dir, focal_dir, "/3pGtoA_freq.txt"), sep = '\t', header = TRUE)
  
  damage_5p_CT$`5pC>T` <- damage_5p_CT$`5pC>T` + focal_5p_file$X5pC.T
  damage_3p_AG$`3pG>A` <- damage_3p_AG$`3pG>A` + focal_3p_file$X3pG.A
  damage_5pCT_ratio <- (sum(damage_5p_CT[1:5,"5pC>T"])) /(sum(damage_5p_CT[6:10,"5pC>T"]))
  damage_5pCT_ratio_5bp <- mean(damage_5p_CT[1:5,"5pC>T"])
  damage_3pAG_ratio <- (sum(damage_3p_AG[1:5,"3pG>A"])) /(sum(damage_3p_AG[6:10,"3pG>A"]))
  damage_3pAG_ratio_5bp <- mean(damage_3p_AG[1:5,"3pG>A"])
  x <- as.data.frame(cbind(taxid_sampleid, damage_5pCT_ratio, damage_5pCT_ratio_5bp, damage_3pAG_ratio, damage_3pAG_ratio_5bp))
  ratio_record <- rbind(ratio_record, x)
  
  y <- focal_5p_file$X5pC.T
  y <- as.data.frame(y)
  colnames(y) <- taxid_sampleid
  all_5p_CT <- cbind(all_5p_CT, y)
  
  z <- focal_3p_file$X3pG.A
  z <- as.data.frame(z)
  colnames(z) <- taxid_sampleid
  all_3p_AG <- cbind(all_3p_AG, z)
  
}

coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio <- ratio_record$damage_5pCT_ratio[match(coverage_depth_dataframe_reads_reduced$taxid_sampleid, ratio_record$taxid_sampleid)]
coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio <- ratio_record$damage_3pAG_ratio[match(coverage_depth_dataframe_reads_reduced$taxid_sampleid, ratio_record$taxid_sampleid)]


damage_5p_CT$`5pC>T` <- damage_5p_CT$`5pC>T` / length(DNA_damage_dirs_reads)
damage_3p_AG$`3pG>A` <- damage_3p_AG$`3pG>A` / length(DNA_damage_dirs_reads)
total_damage <- cbind(damage_5p_CT, damage_3p_AG$`3pG>A`)
colnames(total_damage) <- c("pos" , "5pC>T", "3pG>A")


damage_plot_5p3p <- ggplot(total_damage) +
  #geom_line(aes(x = pos, y = `5pC>T`, color = 'C -> T at 5` terminal'), size = 2) +
  geom_smooth(aes(x = pos, y = `5pC>T`, color = 'C -> T at 5` terminal'), size = 2) +
  #geom_line(aes(x = pos, y = `3pG>A`, color = 'A -> G at 5` terminal'), size = 2) + 
  geom_smooth(aes(x = pos, y = `3pG>A`, color = 'A -> G at 3` terminal'), size = 2) + 
  theme_classic() +
  xlab("Distance from read terminal (bp)") +
  ylab("Rate of C -> T and A -> G transition per bp") +
  scale_color_manual(name = "Nucleotide transition rate:",
                     values = c("darkblue", "darkred")) 
  
plot(damage_plot_5p3p)
#ggsave(paste0(base_dir, "READS_nuc_transition_rate_mean.png"), width = 10, height = 7)


coverage_depth_dataframe_reads_reduced$coverage_scaled_by_ref <- as.numeric(coverage_depth_dataframe_reads_reduced$coverage_scaled_by_ref)
coverage_depth_dataframe_reads_reduced$depth_scaled_by_ref <- as.numeric(coverage_depth_dataframe_reads_reduced$depth_scaled_by_ref)
coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio <- as.numeric(coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio)
coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio <- as.numeric(coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio)


{
coverage_CT_plot <- ggplot(coverage_depth_dataframe_reads_reduced) +
  theme_classic() +
  geom_point(aes(x=coverage_scaled_by_ref, y=damage_5pCT_ratio, color = taxid)) +
  xlab("Average coverage across reference genomes(s)") +
  ylab("Rate of C -> T in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  ylim(c(0, NA)) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_reads_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1)
  
leg <- cowplot::get_legend(coverage_CT_plot)
leg <- ggplotify::as.ggplot(leg)
coverage_CT_plot <- coverage_CT_plot + theme(legend.position = "none")

coverage_AG_plot <- ggplot(coverage_depth_dataframe_reads_reduced) +
  theme_classic() +
  geom_point(aes(x=coverage_scaled_by_ref, y=damage_3pAG_ratio, color = taxid)) +
  xlab("Average coverage across reference genomes(s)") +
  ylab("Rate of A -> G in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  ylim(c(0, NA)) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_reads_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1) +
  scale_x_reverse() +
  scale_y_continuous(position = "right")

coverage_AG_plot <- coverage_AG_plot + theme(legend.position = "none")


max_coverage <- max(max(coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio), 
                    max(coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio))

coverage_CT_plot <- coverage_CT_plot + ylim(c(0,max_coverage))
coverage_AG_plot <- coverage_AG_plot + ylim(c(0,max_coverage))

g <- arrangeGrob(coverage_CT_plot, coverage_AG_plot, leg, nrow= 1) 
plot(g)
#ggsave(file= paste0(base_dir, "READS_rateCT&AG_vs_coverage.png"), g, width = 14, height = 7)
}

{
depth_CT_plot <- ggplot(coverage_depth_dataframe_reads_reduced) +
  theme_classic() +
  geom_point(aes(x=depth_scaled_by_ref, y=damage_5pCT_ratio, color = taxid)) +
  xlab("Average depth across reference genomes(s)") +
  ylab("Rate of C -> T in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  ylim(c(0, NA)) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_reads_reduced$taxid) +
  geom_abline(slope = 0, intercept = 1)
  
leg <- cowplot::get_legend(depth_CT_plot)
leg <- ggplotify::as.ggplot(leg)
depth_CT_plot <- depth_CT_plot + theme(legend.position = "none")

depth_AG_plot <- ggplot(coverage_depth_dataframe_reads_reduced) +
  theme_classic() +
  geom_point(aes(x=depth_scaled_by_ref, y=damage_3pAG_ratio, color = taxid)) +
  xlab("Average depth across reference genomes(s)") +
  ylab("Rate of A -> G in terminal bases positions: 1-5 : 6-10") +
  geom_abline(slope = 0, intercept = 1) +
  ylim(c(0, NA)) +
  scale_color_manual(name = "NCBI taxonomic ID:",
                     values = coverage_depth_dataframe_reads_reduced$taxid) +
  
  geom_abline(slope = 0, intercept = 1) +
  scale_x_reverse() +
  scale_y_continuous(position = "right")

depth_AG_plot <- depth_AG_plot + theme(legend.position = "none")
max_coverage <- max(max(coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio), 
                    max(coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio))

depth_CT_plot <- depth_CT_plot + ylim(c(0,max_coverage))
depth_AG_plot <- depth_AG_plot + ylim(c(0,max_coverage))

g <- arrangeGrob(depth_CT_plot, depth_AG_plot, leg, nrow= 1) 
plot(g)
#ggsave(file= paste0(base_dir, "READS_rateCT&AG_vs_depth.png"), g, width = 14, height = 7)
}

```


```{r}

melted_ratio_record <- melt(ratio_record, id = "taxid_sampleid")
melted_ratio_record_5p <- melted_ratio_record[melted_ratio_record$variable == "damage_5pCT_ratio",]
melted_ratio_record_3p <- melted_ratio_record[melted_ratio_record$variable == "damage_3pAG_ratio",]
melted_ratio_record_both <- rbind(melted_ratio_record_3p, melted_ratio_record_5p)
melted_ratio_record_both <- as.data.frame(melted_ratio_record_both)
melted_ratio_record_both$variable <- as.character(melted_ratio_record_both$variable)
melted_ratio_record_both$variable[melted_ratio_record_both$variable == "damage_5pCT_ratio"] <- "5'"
melted_ratio_record_both$variable[melted_ratio_record_both$variable == "damage_3pAG_ratio"] <- "3'"
colnames(melted_ratio_record_both) <- c("taxid_sampleid", "sense", "ratio")
melted_ratio_record_both$ratio <- as.numeric(melted_ratio_record_both$ratio)

ratio_record_plot <- ggplot(melted_ratio_record_both, aes(x = sense, y = ratio, group = sense, color = sense, fill = sense)) + 
  theme_classic() +
  ylab("Ratio of C->T and A->G transitions\nin terminal bps 0-5 over 6-10") +
  xlab ("Read terminal identity") + 
  theme(legend.position = "none") +
  ylim(c(0,NA)) + 
  geom_abline(slope = 0, intercept = 1)
ratio_record_plot <- ratio_record_plot + geom_violin(trim = FALSE) 
plot(ratio_record_plot)
#ggsave(paste0(base_dir, "READS_ratio_CT_AG.png"), width = 5, height = 7)


all_5p_CT <- cbind(all_5p_CT, damage_5p_CT$`5pC>T`)
all_3p_AG <- cbind(all_3p_AG, damage_3p_AG$`3pG>A`)
names(all_5p_CT)[names(all_5p_CT) == 'damage_5p_CT$`5pC>T`'] <- 'Total 5pC>T'
names(all_3p_AG)[names(all_3p_AG) == 'damage_3p_AG$`3pG>A`'] <- 'Total 3pG>A'


# Remove all columns of zeros in the dataframes:
all_5p_CT <- all_5p_CT[, colSums(all_5p_CT != 0) > 0]
all_3p_AG <- all_3p_AG[, colSums(all_3p_AG != 0) > 0]

# Melts the dataframes, to make plotting the data easier
melted_all_CT <- melt(all_5p_CT, id = "all_5p_CT")
melted_all_AG <- melt(all_3p_AG, id = "all_3p_AG")

# Assign color to each of the sample-taxa that went through DNA damage detection:
set.seed(4)
n_col_df <- melted_all_CT %>%
  group_by(variable) %>%
  dplyr::summarise(n = n())
n_col_df <- as.data.frame(n_col_df)
n_col <- length(n_col_df[,1])  

col_vector2 <- distinctColorPalette(k = n_col, altCol = FALSE, runTsne = FALSE)
n_col_df$color_col <- col_vector2

melted_all_CT$col_column <- n_col_df$color_col[match(melted_all_CT$variable, n_col_df$variable)]

set.seed(4)
n_col_df <- melted_all_AG %>%
  group_by(variable) %>%
  dplyr::summarise(n = n())
n_col_df <- as.data.frame(n_col_df)
n_col <- length(n_col_df[,1])  

col_vector3 <- distinctColorPalette(k = n_col, altCol = FALSE, runTsne = FALSE)
n_col_df$color_col <- col_vector3

melted_all_AG$col_column <- n_col_df$color_col[match(melted_all_AG$variable, n_col_df$variable)]

melted_all_CT <- as.data.frame(melted_all_CT)
names(melted_all_CT)[names(melted_all_CT) == 'variable'] <- 'taxid_sampleid'
names(melted_all_CT)[names(melted_all_CT) == 'all_5p_CT'] <- 'read_pos'
melted_all_CT$col_column[melted_all_CT$taxid == 'Total 5pC>T'] <- '#000000'
melted_all_CT$line_thickness <- 0.1
melted_all_CT$line_thickness[melted_all_CT$taxid == 'Total 5pC>T'] <- 0.5
melted_all_CT$col_column <- c(melted_all_CT$col_column)

melted_all_AG <- as.data.frame(melted_all_AG)
names(melted_all_AG)[names(melted_all_AG) == 'variable'] <- 'taxid_sampleid'
names(melted_all_AG)[names(melted_all_AG) == 'all_3p_AG'] <- 'read_pos'
melted_all_AG$col_column[melted_all_AG$taxid == 'Total 3pG>A'] <- '#000000'
melted_all_AG$line_thickness <- 0.1
melted_all_AG$line_thickness[melted_all_AG$taxid == 'Total 3pG>A'] <- 0.5

melted_all_CT$col_column[melted_all_CT$col_column != "#000000"] <- "skyblue"
melted_all_AG$col_column[melted_all_AG$col_column != "#000000"] <- "skyblue"


melted_all_AG_plot_query <- ggplot(melted_all_AG_query, aes(x = read_pos, y = value)) +
  theme_classic() +
  geom_line(aes(size = line_thickness,  group = taxid_sampleid,  color = I(col_column))) + 
  xlab("Distance from 3' end of read (bp)") +
  ylab("Rate of A -> G transition per bp") +
  scale_size_continuous(name = "Rate of nucleotide\ntranistions:", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) +
  scale_x_reverse() 

{
melted_all_CT_plot <- ggplot(melted_all_CT, aes(x = read_pos, y = value)) +
  theme_classic() +
  geom_line(aes(size = line_thickness,  group = taxid_sampleid,  color = I(col_column))) + 
  xlab("Distance from 5' end of read (bp)") +
  ylab("Rate of C -> T transition per bp") +
  scale_size_continuous(name = "Rate of nucleotide\ntranistions", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) +
  ylim(0, 0.1)

leg <- cowplot::get_legend(melted_all_CT_plot)
leg <- ggplotify::as.ggplot(leg)
melted_all_CT_plot <- melted_all_CT_plot + theme(legend.position = "none")

melted_all_AG_plot <- ggplot(melted_all_AG, aes(x = read_pos, y = value)) +
  theme_classic() +
  geom_line(aes(size = line_thickness,  group = taxid_sampleid,  color = I(col_column))) + 
  xlab("Distance from 3' end of read (bp)") +
  ylab("Rate of A -> G transition per bp") +
  ylim(0, 0.1) +
  scale_size_continuous(name = "Rate of nucleotide\ntranistions", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon"))  + #"Rate of A -> G transition\nper base at 3'"
  scale_x_reverse() +
  scale_y_continuous(limits = c(0,0.1), position = "right")

melted_all_AG_plot <- melted_all_AG_plot + theme(legend.position = "none")
g <- arrangeGrob(melted_all_CT_plot, melted_all_AG_plot, leg, nrow=1) 
plot(g)
#ggsave(file= paste0(base_dir, "READS_CT_AG_transition_perTaxon.png"), g, width = 14, height = 7)
}


all_mtsvreadno <- all_5p_CT[0:1, 2:length(all_5p_CT[1,])]
all_mtsvreadno[1,] <- 0

for(j in 1:length(DNA_damage_dirs_reads)){
  #j <- 3
  focal_dir <- DNA_damage_dirs_reads[j]
  taxid <- sub(".*taxid_", "", focal_dir)
  taxid <- sub("_reads_aligned.*", "", taxid) 
  sampleid <- sub("_taxid_.*", "", focal_dir)
  sampleid <- sub(".*results_", "", sampleid)
  taxid_sampleid <- paste0(taxid, "_", sampleid)
  read_len_file <- paste0(sampleid, "_taxid_", taxid, "_read_lengths.txt")
  #print(read_len_file)
  no_mtsv_reads <- read.table(file = paste0(base_dir, read_len_file), sep = '\t', header = F)
  all_mtsvreadno[1, taxid_sampleid] <- no_mtsv_reads$V1
}

all_mtsvreadno <- all_mtsvreadno[, colSums(all_mtsvreadno != 0) > 0]
total_mtsv_reads <- sum(all_mtsvreadno)



all_5p_CT_2 <- all_5p_CT
damage_5p_CT_2 <- damage_5p_CT
damage_5p_CT_2$`5pC>T` <- 0
mtsv_read_prop <- all_mtsvreadno/total_mtsv_reads
common_taxa <- intersect(colnames(all_5p_CT), colnames(mtsv_read_prop))
mtsv_read_prop <- mtsv_read_prop[, colnames(mtsv_read_prop) == common_taxa]

for(i in 2:(length(all_5p_CT[1,])-1)){
  focal_taxa_sample <- colnames(all_5p_CT[i])
  prop_mtsv <- mtsv_read_prop[1,colnames(mtsv_read_prop) == focal_taxa_sample]
  all_5p_CT_2[,focal_taxa_sample] <- all_5p_CT_2[,focal_taxa_sample] * prop_mtsv
}

all_5p_CT_2$`Total 5pC>T` <- rowSums(all_5p_CT_2[,2:(length(all_5p_CT_2)-1)])
damage_5p_CT_2$`5pC>T` <- all_5p_CT_2$`Total 5pC>T`

all_3p_AG_2 <- all_3p_AG
damage_3p_AG_2 <- damage_3p_AG
damage_3p_AG_2$`3pG>A` <- 0
mtsv_read_prop <- all_mtsvreadno/total_mtsv_reads
common_taxa <- intersect(colnames(all_3p_AG), colnames(mtsv_read_prop))
mtsv_read_prop <- mtsv_read_prop[, colnames(mtsv_read_prop) == common_taxa]

for(i in 2:(length(all_3p_AG[1,])-1)){
  focal_taxa_sample <- colnames(all_3p_AG[i])
  prop_mtsv <- mtsv_read_prop[1,colnames(mtsv_read_prop) == focal_taxa_sample]
  all_3p_AG_2[,focal_taxa_sample] <- all_3p_AG_2[,focal_taxa_sample] * prop_mtsv
}

all_3p_AG_2$`Total 3pG>A` <- rowSums(all_3p_AG_2[,2:(length(all_3p_AG_2)-1)])
damage_3p_AG_2$`3pG>A` <- all_3p_AG_2$`Total 3pG>A`

{
CT_plot_scaled_read_no <- ggplot(damage_5p_CT_2,  aes(x = pos, y = `5pC>T`)) +
  theme_classic() +
  geom_smooth(color = 'dark red', size = 1) + 
  #geom_smooth(color = 'black', size = 0.5, linetype = 'dashed') + 
  #geom_line(color = 'dark red', size = 1) + 
  xlab("Distance from 5' end of read (bp)") +
  ylab("Rate of C -> T transition per bp, scaled by read number") +
  scale_size_continuous(name = "Rate of C -> T transition\nper base at 5'", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) 

AG_plot_scaled_read_no <- ggplot(damage_3p_AG_2,  aes(x = pos, y = `3pG>A`)) +
  theme_classic() +
  geom_smooth(color = 'blue', size = 1) + 
  #geom_smooth(color = 'black', size = 0.5, linetype = 'dashed') + 
  #geom_line(color = 'blue', size = 1) + 
  xlab("Distance from 3' end of read (bp)") +
  ylab("Rate of A -> G transition per bp, scaled by read number") +
  scale_size_continuous(name = "Rate of A -> G transition\nper base at 3'", range = c(0, 2), breaks = c(0.1, 0.5), labels = c("Individual taxon", "Mean per taxon")) +
  scale_x_reverse() +
  scale_y_continuous(position = "right") 

g <- arrangeGrob(CT_plot_scaled_read_no, AG_plot_scaled_read_no, nrow=1) 
plot(g)
#ggsave(file=paste0(base_dir, "READS_CT_AG_transition_perQueryNo.png"), g, width = 14, height = 7)
}
 
saveRDS(damage_5p_CT_2, file = paste0(base_dir, "damage_5p_CT_2.rds"))
saveRDS(damage_3p_AG_2, file = paste0(base_dir, "damage_3p_AG_2.rds"))
saveRDS(damage_5p_CT, file = paste0(base_dir, "damage_5p_CT.rds"))
saveRDS(damage_3p_AG, file = paste0(base_dir, "damage_3p_AG.rds"))

```



```{r}

max_ratio <- max(max(coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio), max(coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio), max(coverage_depth_dataframe_queries_reduced$damage_3pAG_ratio), max(coverage_depth_dataframe_queries_reduced$damage_5pCT_ratio))

{
READS_depth_coverage_5p_relationship <- ggplot(data = coverage_depth_dataframe_reads_reduced, aes(x = coverage_scaled_by_ref, y = depth_scaled_by_ref)) +
  theme_classic() +
  geom_point(aes(color = coverage_depth_dataframe_reads_reduced$damage_5pCT_ratio), size = 2, alpha = 1) +
  xlab("Percentage coverage (reads)") +
  ylab("Percentage depth (reads)") +
  xlim(c(0,25)) +
  ylim(c(0,2)) +
  theme(axis.text=element_text(size=12)) +
  scale_color_gradientn(colours = c("darkred","white","royalblue", "royalblue4","darkblue", "black"), 
                        values = c(0,  max(damage_3pAG_ratio)/(max_ratio*(3/5)), 1),
                        guide = "colorbar", limits=c(0, max_ratio),
                        name = "Cumulative nucleotide transition rate\nin terminal bases 1-5 over 6-10")
  
leg <- cowplot::get_legend(READS_depth_coverage_5p_relationship)
leg <- ggplotify::as.ggplot(leg)
READS_depth_coverage_5p_relationship <- READS_depth_coverage_5p_relationship + theme(legend.position = "none")

READS_depth_coverage_3p_relationship <- ggplot(data = coverage_depth_dataframe_reads_reduced, aes(x = coverage_scaled_by_ref, y = depth_scaled_by_ref)) +
  theme_classic() +
  geom_point(aes(color = coverage_depth_dataframe_reads_reduced$damage_3pAG_ratio), size = 2, alpha = 1) +
  xlab("Percentage coverage (queries)") +
  ylab("Percentage depth (queries)") +
  xlim(c(0,25)) +
  ylim(c(0,2)) +
  theme(axis.text=element_text(size=12)) +
  scale_color_gradientn(colours = c("darkred","white","royalblue", "royalblue4","darkblue", "black"), 
                        values = c(0,  max(damage_3pAG_ratio)/(max_ratio*(3/5)), 1),
                        guide = "colorbar", limits=c(0, max_ratio),
                        name = "Cumulative nucleotide transition rate\nin terminal bases 1-5 over 6-10")

READS_depth_coverage_3p_relationship <- READS_depth_coverage_3p_relationship + 
                                        theme(legend.position = "none",
                                              axis.title.y=element_blank())

g <- arrangeGrob(READS_depth_coverage_5p_relationship, READS_depth_coverage_3p_relationship, leg, nrow=1) 
plot(g)
#ggsave(file = paste0(base_dir, "READS_depth_coverage_5p3p_relationship.png"), g, width = 14, height = 7)

}


```

